Mathlib/Algebra/Order/Incidence.lean

```diff
@@ -0,0 +1,645 @@
+  zero := 0
+
+instance instRing [Preorder α] [LocallyFiniteOrder α] [DecidableEq α] [Ring 𝕜] :
+    Ring (IncidenceAlgebra 𝕜 α) where
+  __ := instSemiring
+  __ := instAddGroup
+
+/-! ### Scalar multiplication between incidence algebras -/
+
+section Smul
+variable [Preorder α] [LocallyFiniteOrder α] [AddCommMonoid 𝕜] [AddCommMonoid 𝕝] [SMul 𝕜 𝕝]
+
+instance instSMul : SMul (IncidenceAlgebra 𝕜 α) (IncidenceAlgebra 𝕝 α) :=
+  ⟨fun f g ↦
+    ⟨fun a b ↦ ∑ x ∈ Icc a b, f a x • g x b, fun a b h ↦ by dsimp; rw [Icc_eq_empty h, sum_empty]⟩⟩
+
+@[simp]
+lemma smul_apply (f : IncidenceAlgebra 𝕜 α) (g : IncidenceAlgebra 𝕝 α) (a b : α) :
+    (f • g) a b = ∑ x ∈ Icc a b, f a x • g x b :=
+  rfl
+
+end Smul
+
+instance instIsScalarTower [Preorder α] [LocallyFiniteOrder α] [AddCommMonoid 𝕜] [Monoid 𝕜]
+    [Semiring 𝕝] [AddCommMonoid 𝕞] [SMul 𝕜 𝕝] [Module 𝕝 𝕞] [DistribMulAction 𝕜 𝕞]
+    [IsScalarTower 𝕜 𝕝 𝕞] :
+    IsScalarTower (IncidenceAlgebra 𝕜 α) (IncidenceAlgebra 𝕝 α) (IncidenceAlgebra 𝕞 α) where
+  smul_assoc f g h := by
+    ext a b
+    simp only [smul_apply, sum_smul, smul_sum, sum_sigma']
+    apply sum_nbij' (fun ⟨a, b⟩ ↦ ⟨b, a⟩) (fun ⟨a, b⟩ ↦ ⟨b, a⟩) <;> aesop (add unsafe le_trans)
+
+instance [Preorder α] [LocallyFiniteOrder α] [DecidableEq α] [Semiring 𝕜] [Semiring 𝕝]
+    [Module 𝕜 𝕝] : Module (IncidenceAlgebra 𝕜 α) (IncidenceAlgebra 𝕝 α) where
+  smul := (· • ·)
+  one_smul f := by ext a b hab; simp [ite_smul, hab]
+  mul_smul f g h := by convert smul_assoc _ _ _; ext a b; rfl; infer_instance
+  smul_add f g h := by ext; exact Eq.trans (sum_congr rfl fun x _ ↦ smul_add _ _ _) sum_add_distrib
+  add_smul f g h := by ext; exact Eq.trans (sum_congr rfl fun x _ ↦ add_smul _ _ _) sum_add_distrib
+  zero_smul f := by ext; exact sum_eq_zero fun x _ ↦ zero_smul _ _
+  smul_zero f := by ext; exact sum_eq_zero fun x _ ↦ smul_zero _
+
+instance smulWithZeroRight [Zero 𝕜] [Zero 𝕝] [SMulWithZero 𝕜 𝕝] [LE α] :
+    SMulWithZero 𝕜 (IncidenceAlgebra 𝕝 α) :=
+  DFunLike.coe_injective.smulWithZero ⟨((⇑) : IncidenceAlgebra 𝕝 α → α → α → 𝕝), coe_zero⟩
+    coe_constSMul
+
+instance moduleRight [Preorder α] [Semiring 𝕜] [AddCommMonoid 𝕝] [Module 𝕜 𝕝] :
+    Module 𝕜 (IncidenceAlgebra 𝕝 α) :=
+  DFunLike.coe_injective.module _ ⟨⟨((⇑) : IncidenceAlgebra 𝕝 α → α → α → 𝕝), coe_zero⟩, coe_add⟩
+    coe_constSMul
+
+instance algebraRight [PartialOrder α] [LocallyFiniteOrder α] [DecidableEq α] [CommSemiring 𝕜]
+    [CommSemiring 𝕝] [Algebra 𝕜 𝕝] : Algebra 𝕜 (IncidenceAlgebra 𝕝 α) where
+  toFun c := algebraMap 𝕜 𝕝 c • (1 : IncidenceAlgebra 𝕝 α)
+  map_one' := by
+    ext; simp only [mul_boole, one_apply, Algebra.id.smul_eq_mul, constSMul_apply, map_one]
+  map_mul' c d := by
+      ext a b
+      obtain rfl | h := eq_or_ne a b
+      · simp only [one_apply, Algebra.id.smul_eq_mul, mul_apply, Algebra.mul_smul_comm,
+          boole_smul, constSMul_apply, ← ite_and, map_mul, Algebra.smul_mul_assoc,
+          if_pos rfl, eq_comm, and_self_iff, Icc_self]
+        simp
```

Comment:
This one looks like you can move more bits into the `simp`!

Reply:
Heartbeats go from 22k to timeout when I do that